compiler/AST/ParamForLoop.cpp

```diff
@@ -61,14 +62,42 @@ BlockStmt* ParamForLoop::buildParamForLoop(VarSymbol* indexVar,
     stride = new SymExpr(new_IntSymbol(1));
   }
 
-  if (call && call->isNamed("chpl_build_bounded_range"))
+  if (call && call->isNamed("#"))
   {
-    low    = call->get(1)->remove();
-    high   = call->get(1)->remove();
+    count = new CallExpr("chpl_compute_count_param_loop", call->get(2)->remove());
+    call        = toCallExpr(call->get(1));
+  }
+
+  if (count != NULL && call && call->isNamed("chpl_build_low_bounded_range"))
+  {
+    low  = call->get(1)->remove();
+    high = new CallExpr("chpl_high_bound_count_for_param_loop", low->copy(), count);
+  }
+  else if (count != NULL && call && call->isNamed("chpl_build_high_bounded_range"))
+  {
+    high = call->get(1)->remove();
+    low  = new CallExpr("chpl_low_bound_count_for_param_loop", high->copy(), count);
+  }
+  else if (call && call->isNamed("chpl_build_bounded_range"))
+  {
+    if(count != NULL)
+    {
+      Expr* temp_low = call->get(1)->remove();
+      Expr* temp_high = call->get(1)->remove();
+
+      // It is necessary that low is calculated first, because it also applies check for bound size.
+      low = new CallExpr("chpl_bounded_count_for_param_loop_low", temp_low->copy(), temp_high->copy(), count->copy());
+      high = new CallExpr("chpl_bounded_count_for_param_loop_high", temp_low->copy(), temp_high->copy(), count);
+    }
+    else
+    {
+      low    = call->get(1)->remove();
+      high   = call->get(1)->remove();
+    }
```

Comment:
I think this latest refactoring cleans this code up a lot, thanks!  I'd like to propose one more refactoring to put what I'd consider the most common / important cases first (which is generally a good practice to help people reading the code to focus in on the simple / straightforward cases first and then the more exceptional ones later):

```chapel
if (call && call->isNamed("chpl_build_bounded_range")) {
  if (count == NULL) {
    // handle low..high
    ...
  } else {
    // handle low..high#count
    ...
  }
} else if (call && count != NULL) {
  // handle lo..#count
  if (call->isNamed("chpl_build_low_bounded_range")) {
    ...
  } else if (call->isNamed("chpl_build_high_bounded_range") {
    ...
  }
} else {
}
```

The main downside to this approach (that I can think of) is that the error cases are now spread across multiple conditionals again.  For that reason, I think I'd switch from using an `else` for the error case to having a separate conditional after this big one that simply says:

```chapel
if (low == NULL || high == NULL) {
  USR_FATAL(range, "param for loops currently only support range expressions with well defined param integral bounds");
}
```

This would also permit us to hoist the `if (call)` to the very to of the conditional I sketched above to avoid having to check it in multiple places:

```chapel
if (call) {
  if (call->isNamed("chpl_build_bounded_range")) {
    ...
  } else if (count != NULL) {
   if (call->isNamed(...)) {
```
etc.

What do you think?

Reply:
Sure sir. This is nice. I will update it now.